In [10]:
# AAE 722 Homework 1 Charles Voigt
#Housekeeping
import numpy as np
import pandas as pd
from matplotlib.pyplot import subplots
import statsmodels.api as sm
#!pip install ISLP

from statsmodels.stats.outliers_influence \
    import variance_inflation_factor as VIF
from statsmodels.stats.anova import anova_lm

from ISLP import load_data
from ISLP.models import (ModelSpec as MS, summarize, poly)


In [3]:
#Question 1.1 - Replicating the Regression Table
url = "https://www.statlearning.com/s/Advertising.csv"
Advertising = pd.read_csv(url)

#Here I will create the regression

X = pd.DataFrame({'intercept': np.ones(Advertising.shape[0]),
                  'TV': Advertising['TV'],
                  'radio': Advertising['radio'],
                  'newspaper': Advertising['newspaper']})
X[:4]
y = Advertising['sales']

model = sm.OLS(y, X).fit()
summarize(model)

coef  std err       t  P>|t|
intercept  2.9389    0.312   9.422   0.00
TV         0.0458    0.001  32.809   0.00
radio      0.1885    0.009  21.893   0.00
newspaper -0.0010    0.006  -0.177   0.86


### Question 1.2

Question: For each predictor (TV, radio, and newspaper), describe the null hypothesis that corresponds to the p-values given in the regression table.

TV: <br>
   H<sub>0</sub>: $\beta$<sub>TV</sub> = 0 <br>
   H<sub>1</sub>: $\beta$<sub>TV</sub> != 0<br>
   reject if p value > 0.05 <br>\
radio: <br>
   H<sub>0</sub>: $\beta$<sub>radio</sub> = 0 <br>
   H<sub>1</sub>: $\beta$<sub>radio</sub> != 0<br>
   reject if p value > 0.05 <br>\
newspaper: <br>
   H<sub>0</sub>: $\beta$<sub>newspaper</sub> = 0 <br>
   H<sub>1</sub>: $\beta$<sub>newspaper</sub> != 0<br>
   reject if p value > 0.05 <br>

Explain what the null hypothesis means in the context of the data, not just in terms
of the coefficients.

The null hypothesis (H<sub>0</sub>) describes a scenario in which TV (or radio, or newspaper) have no association with sales. In the context of the data provided by the table, this means that TV and radio advertising have some association (0.046 and 0.189) with sales because the p value > 0.05 (assuming a .05 significance level). Newspaper advertising has a p value of 0.8599. Therefore, we fail to reject the null hypothesis that $\beta$<sub>newspaper</sub> = 0, and therefore conclude that newspaper advertising has no association with sales.


### Question 1.3

Question: Based on the p-values from the regression output, explain what conclusions
you can draw about the relationship between sales and the predictors (TV, radio,
newspaper).

As discussed above, TV and radio advertising were found to be statistically significant at the 5% level (p value <0.05). Therefore, we can conclude that the betas for TV and radio aptly describe the relationship between these mediums of advertising and sales. If sales are measured in # of units sold, and advertising is measured in $ spent (thousands), the relationship for TV is then described as: *For every thousand dollars spent on TV advertising, sales will increase by 0.046 units.* Similarly, for radio advertising, the relationship is described as: *For every thousand dollars spent on radio advertising, sales will increase by 0.189 units.* <br>

For newspaper advertising, it was found that the p value was 0.8599, (>0.05), therefore we failed to reject the null hypothesis and concluded that the beta for newspaper advertising was 0. This means that newspaper advertising has no measurable relationship with sales.


### Question 2

The K Nearest Neighbors Classifier is a method of supervised learning which uses probability to assign a 'class', or group to an observation based on the groups which the K number of the observations 'nearest neighbors', or closest matches belong to.

The K Nearest Neighbors Regression method is similar to the classifier. It uses K nearest neighbors to make a prediction about a new observation, but instead of placing the observation into a group (classification), it takes the average value of the K nearest neighbors to predict the new observation. This method is useful for predicting a continuous value, as opposed to a discrete classification.

### Question 3

#### (a)
The RSS for the cubic regression will be less than the linear RSS. The RSS calculation will choose the values of the betas to minimize RSS. Because n=100 (not infinite), some of the irreducible error in the model will be captured by the cubic model in $\beta$<sub>2</sub> and $\beta$<sub>3</sub>. Therefore, the RSS will be lower, but not necessarily indicative of a better model specification.

#### (b)
Here, the RSS will be the opposite: larger for the cubic regression than the linear. This is because the cubic model will be overfitted, and attempt to predict the signal based on noise. 

#### (c)

In this case, the RSS for the cubic regression will be smaller than the linear regression. Some of the nonlinearity in the true model will be captured by the cubic regression, whereas the linear regression will not capture this relationship. Thus, in the training data, the cubic regression will yield a better fit and therefore a smaller RSS.


(d)

Using the test data, it is impossible to determine which RSS will be lower. This is because the true model is unknown. If it is very close to linear, the linear regression model will yield a lower RSS and the cubic regression will be overfit. If the true model is highly nonlinear, then the cubic regression will be a better estimate and the linear regression will be underfit.

In [11]:
# Question 4 (a)

rng = np.random.default_rng(1)
x = rng.normal(size=100)
y = 2 * x + rng.normal(size=100)

model = sm.OLS(y, x).fit()
MS(intercept=False)
summarize(model)

coef  std err       t  P>|t|
x1  1.9762    0.117  16.898    0.0

#### Question 4 (a) Discussion

The coefficient of $\beta$<sub>1</sub> is 1.9762, with a standard error of 0.117 and t statistic of 16.898 (p value = 0). At the 5% level, we reject the null hypothesis that $\beta$ = 0, and conclude that a one unit increase in x is associated with a 1.9762 unit increase in y.

In [12]:
# Question 4 (b)

model =sm.OLS(x, y).fit()
MS(intercept=False)
summarize(model)


coef  std err       t  P>|t|
x1  0.3757    0.022  16.898    0.0

#### Question 4 (b) Discussion

The coefficient of $\beta$<sub>1</sub> is 0.3757, with a standard error of 0.022, t statistic of 16.898, and p value of 0.0. At the 5% level, we reject the null hypothesis that $\beta$<sub>1</sub> = 0, and conclude that a one unit increase in y is associated with a 0.3757 increase in x.

#### Question 4 (c) Discussion

While the t statistics of the two regressions are identical (indicating that the correlation is the same), the beta coefficients are not the same nor inverses (reciprocals). This is because the standard errors for the variables are different, and the variances in the regression causes the slopes of the coefficients to differ.

#### Question 4 (d)

SE($\hat{\beta}$) = $\frac{\sqrt{\sum(y_i^2-2x_iy_i\hat{\beta}+\hat{\beta}^2x_i^2)}}{\sqrt{(n-1)\sum(x_i^2)}}$\
\
 = $\frac{\sqrt{\sum(y_i^2-2\frac{(x_iy_i)^2}{x_i^2}+\frac{(x_iy_i)^2}{x_i^2})}}{\sqrt{(n-1)\sum(x_i^2)}}$ \
 \
 = $\frac{\sqrt{\sum(y_i^2-\frac{(x_iy_i)^2}{x_i^2})}}{\sqrt{(n-1)\sum(x_i^2)}}$

$\hat{\beta} = \frac{\sum_{i=1}^{n} x_i y_i}{\sum_{i=1}^{n} x_i^2}$ \
\
$\hat{\beta}$/ SE($\hat{\beta}$) = $\frac{{\sum_{i=1}^{n} x_i^2}}{\sum_{i=1}^{n} x_i y_i}$ $\frac{\sqrt{\sum(y_i^2-\frac{(x_iy_i)^2}{x_i^2})}}{\sqrt{(n-1)\sum(x_i^2)}}$\
\
= $\frac{\sqrt{(n-1)}\sum(x_iy_i)}{\sqrt{(\sum{x_i^2})\sum(y_i^2-\frac{(x_iy_i)^2}{x_i^2})}}$\
\
= $\frac{\sqrt{(n-1)}\sum(x_iy_i)}{\sqrt{(\sum{x_i^2})(\sum{y_i^2})-(\sum{x_iy_i})^2}}$

In [18]:
# Question 4 (d)
rng = np.random.default_rng(1)
x = rng.normal(size=100)
y = 2 * x + rng.normal(size=100)

n = len(x)
xy = np.sum(x * y)
x2 = np.sum(x ** 2)
y2 = np.sum(y ** 2)

(np.sqrt(n-1) * xy) / np.sqrt((x2 * y2) - (xy**2))

16.898417063035094

#### Question 4 (e)

The t statistic will be the same for both regressions because it is a measure of correlation between the two variables, which will be identical regardless of which variable is the dependent variable. Part d illustrates this in the final equation by showing algebraically that x and y can be swapped without yielding a different result.

In [25]:
# Question 4 (f)

rng = np.random.default_rng(1)
x = rng.normal(size=100)
y = 2 * x + rng.normal(size=100)

X1 = sm.add_constant(x)

model = sm.OLS(y, X1).fit()
MS(intercept=True)

X2 = sm.add_constant(y)
model2 = sm.OLS(x, X2).fit()
MS(intercept=True)

summarize(model), summarize(model2)


(         coef  std err       t  P>|t|
 const -0.0760    0.101  -0.756  0.451
 x1     1.9686    0.118  16.734  0.000,
          coef  std err       t  P>|t|
 const  0.0095    0.044   0.216  0.829
 x1     0.3763    0.022  16.734  0.000)